server/webapp/WEB-INF/rails.new/spec/javascripts/websocket_wrapper_spec.js

```diff
@@ -0,0 +1,362 @@
+
+    assertEquals(false, mySinon.isCalled);
+    assertEquals(0, mySinon.callCount);
+
+    myWebsocket.ws.onopen();
+    myWebsocket.ws.onerror({data: "Boom!"});
+
+    assertEquals(true, mySinon.isCalled);
+    assertEquals(1, mySinon.callCount);
+  });
+
+  it("test_websocket_wrapper_should_emit_initial_connection_failed_event_when_an_error_occurs_while_websocket_handshake", function () {
+    var options       = {};
+    options.Websocket = MockedWebsocket;
+    var myWebsocket   = new WebSocketWrapper(options);
+    myWebsocket.on("initialConnectFailed", mySinon.stub);
+
+    assertEquals(false, mySinon.isCalled);
+    assertEquals(0, mySinon.callCount);
+
+    myWebsocket.ws.onerror({data: "Boom!"});
+
+    assertEquals(true, mySinon.isCalled);
+    assertEquals(1, mySinon.callCount);
+  });
+
+  it("test_websocket_wrapper_should_retry_on_error_event", function () {
+    var options                    = {};
+    options.Websocket              = MockedWebsocket;
+    options.indefiniteRetry        = true;
+    var myWebsocket                = new WebSocketWrapper(options);
+    WebSocketWrapper.TIMEOUT_START = 0;
+    myWebsocket.on("error", function () {
+    });
+
+    myWebsocket.on("beforeInitialize", mySinon.stub);
+
+    assertEquals(false, mySinon.isCalled);
+    assertEquals(0, mySinon.callCount);
+
+    myWebsocket.ws.onopen();
+    myWebsocket.ws.onerror({data: "Boom!"});
+    setTimeout(function () {
+      assertEquals(true, mySinon.isCalled);
+      assertEquals(1, mySinon.callCount);
+    }, 10);
+  });
+
+  it("test_websocket_wrapper_should_not_retry_on_error_event_when_indefiniteRetry_is_false", function () {
+    var options                    = {};
+    options.Websocket              = MockedWebsocket;
+    options.indefiniteRetry        = false;
+    var myWebsocket                = new WebSocketWrapper(options);
+    WebSocketWrapper.TIMEOUT_START = 0;
+    myWebsocket.on("error", function () {
+    });
+
+    myWebsocket.on("beforeInitialize", mySinon.stub);
+
+    assertEquals(false, mySinon.isCalled);
+    assertEquals(0, mySinon.callCount);
+
+    myWebsocket.ws.onopen();
+    myWebsocket.ws.onerror({data: "Boom!"});
+
+    setTimeout(function () {
+      assertEquals(false, mySinon.isCalled);
+      assertEquals(0, mySinon.callCount);
+    }, 10);
+  });
+
+  it("test_websocket_wrapper_should_not_retry_on_error_event_when_initial_connection_fails", function () {
+    var options                    = {};
+    options.Websocket              = MockedWebsocket;
+    options.indefiniteRetry        = true;
+    var myWebsocket                = new WebSocketWrapper(options);
+    WebSocketWrapper.TIMEOUT_START = 0;
+    myWebsocket.on("error", function () {
+    });
+
+    myWebsocket.on("beforeInitialize", mySinon.stub);
+
+    assertEquals(false, mySinon.isCalled);
+    assertEquals(0, mySinon.callCount);
+
+    myWebsocket.ws.onerror({data: "Boom!"});
+
+    setTimeout(function () {
```

Comment:
This is a bad idea — because the callback is invoked after the test finishes, exceptions raised will likely be marked on a wrong test. Worse, if the callback is invoked after all tests finish execution, the exception will never be recognized.

Check https://jasmine.github.io/2.3/introduction#section-Asynchronous_Support